src/mpid/ch4/generic/am/mpidig_am_part_callbacks.c

```diff
@@ -31,7 +31,7 @@ static int part_send_data_target_cmpl_cb(MPIR_Request * rreq)
     MPID_Request_complete(rreq);
 
     /* Reset part_rreq status to inactive */
-    MPL_atomic_store_int(&MPIDIG_PART_REQUEST(part_rreq, status), MPIDIG_PART_RREQ_INACTIVE);
+    MPL_atomic_fetch_sub_int(&MPIDIG_PART_REQUEST(part_rreq, status), 1);
 
```

Comment:
I'd suggest aways use macro instead of magic numbers. I also put the comment in the next commit.

Reply:
I disagree here. The semantics with the "status" here is a numeric semantics -- it starts as `0`, add `1` for each necessary condition is met; it checks agast a final value to decide whether "all" conditions are met. In this semantics, `0` is not magic. `0` means `0` as every numeric context everywhere. On the other hand, the names made the simple numeric context cryptic. For example, `UNSET + 1 -> PARTIAL_ACTIVATED`. `PARTIAL_ACTIVATED + 1 -> REQ_CTS`. It makes `0 + 1 = 1` and `1 + 1 = 2` obscure. In fact, I am leaning toward removing all names. Code like
```
if (status == 2) {
    /* Both conditions met, can send data */
    issue_data
}
```
seems very straight forward. On the other hand,
```
if (status == MPIDIG_PART_REQ_CTS) {
    issue_data
}
```
seems mysterious.

Reply:
I guess that we cannot converge. Maybe you can ping the other reviewer to co-approve this commit?

Reply:
So when I was thinking about how to implement a CTS scheme in the netmod, I initially drew up local and remote epoch counters that would be incremented for each `MPI_Start`, and the receiver process could potentially use a remote atomic to update the remote epoch on the sender process. Data transfer would only start when `local == remote`. From what I can tell, this code implements essentially the same just with a single var, which does make it a little hard to understand for the reader. Both the old (buggy) version and this fixed version.

IMO, this version is better but still a bit difficult to understand. What do you think of having separate local and remote values?

Reply:
I think it makes sense. But probably a better name/prefix is `sender_` and `receiver_` rather than `local_` and `remote_`.  How about let me drop the last commit, so @raffenet you could eventually replace it with your new versions.

Reply:
Sure, I can give that a try.

Reply:
@hzhou can we replace direct use of `1` or `2`  with `(MPIDIG_PART_REQ_CTS-MPIDIG_PART_SREQ_INACTIVE)` or `(MPIDIG_PART_REQ_CTS-MPIDIG_PART_RREQ_INACTIVE)`? Without the additional comment in your other commit, the `1` and `2` are magic. If we want to keep this fix PR simple, I'd suggest we use the current MACROs.

Reply:
What does *subtract* `MPIDIG_PART_REQ_CTS-MPIDIG_PART_SREQ_INACTIVE` mean? It doesn't make sense to me.

Reply:
It reads as reduce the status from CTS to INACTIVE.

Reply:
No. `a - 1` is not reduce `a` to `1`.

Reply:
@minsii I guess you haven't really understood what the bug was. Your assumption that `status` is `CTS` (2) is incorrect, and that was the bug. The `status` could be `3` due to the next `CTS` arrive before the previous send completes.

Reply:
Right. This is why I think having separate send and recv epoch vars would be easier to understand.

Reply:
To be clear, I think the fix in this PR is fine to merge for now.

Reply:
I see the key issue here is that, the "remote-started" condition is updated by the remote process, thus it may be out-of-order with the update of the local conditions. The current code covers the cases of (1) local-completed -> local-started -> remote-started and (2) local-completed -> remote-started -> local-started. But it does not cover (3)  remote-started -> local-completed -> local-started. The (3) case is fixed in this PR. I think there should be no other cases.

I agreed that having separate local and remote vars  (or send/recv) gives a better reading. But I guess the same out-of-order issue may occur, as the remote var is updated by am.

Reply:
The key concern of the second commit is that, the cntr/stat logic is a bit complex (it was a fault of my original code) thus we want to keep the logic definition in one place. The original code defines status macros in a single place as the logic definition. But now it splits into two parts -- the part uses the status, and the part directly uses -1 or -2 numbers.

I hope the logic can be still defined in a single place, thus don't make the code more complex. That was why suggested to still use status macros instead of -1/-2. Using `(CTS-INACTIVE)` might be misleading. How about define `NUM_CTS_CONDITIONS`? Besides, these `INACTIVE` status macros should be removed as they are unused now.

Reply:
Alright, I'll name them.

Reply:
@minsii I added the last commit to address your concern.